{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Regression Analysis of Sabin data "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This analsys has been done in order to try and determine the context effects\n",
    "of different context on the substitution probability\n",
    "The input data is a csv files of the different K-mers and the $log[Pr(A\\to B|S]$ of substitution rates of each of the passeges.\n",
    "\n",
    "It includes 5 stages:\n",
    "\n",
    "1. Feature extraction- creating a dummy variable for the categorical feature of context in a certain position\n",
    "2. Feature selection- selecting the most important features using the randomized lasso regression method\n",
    "3. Signifcance and effect test- after constructing the regression formula we are applying it in the regression to see if the effect of the motiff is significant also find the $\\beta$ coeffs \n",
    "\n",
    "4. summarize the results across the different mutation types and passeges\n",
    "5. Motif analysis - mining and displaying data about the different motifs to support decision about biological meaning of the motifs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "import itertools\n",
    "import numpy as np\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from treeinterpreter import treeinterpreter as ti\n",
    "from scipy.stats import kendalltau,spearmanr\n",
    "import statsmodels.formula.api as sm\n",
    "from sklearn.linear_model import RandomizedLasso\n",
    "import os\n",
    "import glob\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import matplotlib.pyplot as plt\n",
    "global MUTATIONS\n",
    "MUTATIONS=['AC', 'GT', 'AG', 'CA', 'CG', 'GC', 'AT', 'GA', 'TG', 'CT', 'TC', 'TA']\n",
    "global NUCS\n",
    "NUCS=['A','C','G','T']\n",
    "plt.style.use('ggplot')\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Seq import reverse_complement, transcribe, back_transcribe, translate\n",
    "from Bio import SeqIO\n",
    "from scipy.stats import hypergeom\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample of a data file in Passege 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>mutationType</th>\n",
       "      <th>P1</th>\n",
       "      <th>P2</th>\n",
       "      <th>P3</th>\n",
       "      <th>P4</th>\n",
       "      <th>P5</th>\n",
       "      <th>prob</th>\n",
       "      <th>createCpG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTAA</td>\n",
       "      <td>CT</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>-7.819754</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTATC</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-20.723266</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATAGG</td>\n",
       "      <td>GA</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>-20.723266</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    kmer mutationType P1 P2 P3 P4 P5       prob createCpG\n",
       "0  ACTAA           CT  A  C  C  A  A  -7.819754     False\n",
       "1  TTATC           TA  T  T  T  T  C -20.723266     False\n",
       "2  ATAGG           GA  A  T  G  G  G -20.723266     False"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data=pd.read_csv(r'C:\\Users\\Guyling1\\ContextProject\\regressionCSVOneFileSABINK5P7.csv')\n",
    "data.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.Feature Extraction\n",
    "\n",
    "Defined a function that adds dummy variables of context dependence up to 3rd degree \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def generateFeature(k):\n",
    "    for j in range(1,k+1):\n",
    "        for feature in itertools.imap(''.join,itertools.combinations('1245',j)):\n",
    "            yield feature\n",
    "\n",
    "def functionGenerator(positions,x):\n",
    "    res=''\n",
    "    for position in positions:\n",
    "        res+=x['P{}'.format(position)]\n",
    "    return res\n",
    "\n",
    "def createFeaturePrefix(featureList):\n",
    "    res=''\n",
    "    for feature in featureList:\n",
    "        res+=\"P{}_\".format(feature)\n",
    "    return res\n",
    "\n",
    "def getDummyVarPositionForK(data,k):\n",
    "    for feature in generateFeature(k):\n",
    "        featureList=[int(x) for x in list(feature)]\n",
    "        concatFeature=data.apply(lambda x: functionGenerator(featureList,x),axis=1)\n",
    "        dummies=pd.get_dummies(concatFeature,prefix=createFeaturePrefix(featureList))\n",
    "        #print createFeaturePrefix(featureList) \n",
    "        data=pd.concat([data,dummies],axis=1)\n",
    "    return data\n",
    "\n",
    "def cleanDataMatrix(data,withNull=False):\n",
    "    for i in range(1,6):\n",
    "        data=data.drop('P{}'.format(i),axis=1)\n",
    "    data=data.drop('createCpG',axis=1)\n",
    "    if not(withNull):\n",
    "        data=data[data['prob']>-20]\n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample of the data after adding the dummy var features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kmer</th>\n",
       "      <th>mutationType</th>\n",
       "      <th>prob</th>\n",
       "      <th>P1__A</th>\n",
       "      <th>P1__C</th>\n",
       "      <th>P1__G</th>\n",
       "      <th>P1__T</th>\n",
       "      <th>P2__A</th>\n",
       "      <th>P2__C</th>\n",
       "      <th>P2__G</th>\n",
       "      <th>...</th>\n",
       "      <th>P2_P4_P5__TCG</th>\n",
       "      <th>P2_P4_P5__TCT</th>\n",
       "      <th>P2_P4_P5__TGA</th>\n",
       "      <th>P2_P4_P5__TGC</th>\n",
       "      <th>P2_P4_P5__TGG</th>\n",
       "      <th>P2_P4_P5__TGT</th>\n",
       "      <th>P2_P4_P5__TTA</th>\n",
       "      <th>P2_P4_P5__TTC</th>\n",
       "      <th>P2_P4_P5__TTG</th>\n",
       "      <th>P2_P4_P5__TTT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTAA</td>\n",
       "      <td>CT</td>\n",
       "      <td>-7.819754</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCACT</td>\n",
       "      <td>TA</td>\n",
       "      <td>-10.486637</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TTTAT</td>\n",
       "      <td>CT</td>\n",
       "      <td>-8.627929</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 371 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    kmer mutationType       prob  P1__A  P1__C  P1__G  P1__T  P2__A  P2__C  \\\n",
       "0  ACTAA           CT  -7.819754    1.0    0.0    0.0    0.0    0.0    1.0   \n",
       "3  GCACT           TA -10.486637    0.0    0.0    1.0    0.0    0.0    1.0   \n",
       "9  TTTAT           CT  -8.627929    0.0    0.0    0.0    1.0    0.0    0.0   \n",
       "\n",
       "   P2__G      ...        P2_P4_P5__TCG  P2_P4_P5__TCT  P2_P4_P5__TGA  \\\n",
       "0    0.0      ...                  0.0            0.0            0.0   \n",
       "3    0.0      ...                  0.0            0.0            0.0   \n",
       "9    0.0      ...                  0.0            0.0            0.0   \n",
       "\n",
       "   P2_P4_P5__TGC  P2_P4_P5__TGG  P2_P4_P5__TGT  P2_P4_P5__TTA  P2_P4_P5__TTC  \\\n",
       "0            0.0            0.0            0.0            0.0            0.0   \n",
       "3            0.0            0.0            0.0            0.0            0.0   \n",
       "9            0.0            0.0            0.0            0.0            0.0   \n",
       "\n",
       "   P2_P4_P5__TTG  P2_P4_P5__TTT  \n",
       "0            0.0            0.0  \n",
       "3            0.0            0.0  \n",
       "9            0.0            0.0  \n",
       "\n",
       "[3 rows x 371 columns]"
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data=getDummyVarPositionForK(data,3)\n",
    "data=cleanDataMatrix(data)\n",
    "data.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After adding the dummy variables we want to create a feature matrix $X$ and predicted value vector $\\bar{y}$  so that we can easily use it in the regression .\n",
    "\n",
    "WithNull parameter chooses if we want to include unobserved mutations with their defult value of $10^{-9}$\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def createCovarsAndProbVectorForMutationType(data,mutType):\n",
    "    data=data[data['mutationType']==mutType]\n",
    "    probVector=data.prob\n",
    "    covars=data.drop('prob',axis=1).drop('kmer',axis=1).drop('mutationType',axis=1)\n",
    "    return covars,probVector,data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.Feature Selection\n",
    "\n",
    "Because we have a very large number of features and these features are also dependent (for example there is an obvious dependecy between P1=A and P1=A,P2=C) we use <a href=\"http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RandomizedLasso.html\">Randomized lasso regression</a> which a randomized  $L-1$ norm based method of regression. We will only take in to account features that were used in over $ratio=0.2$ of the cases. This was done in order to solve problems with the forward feature selection methods that did not preform well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def randomLassoFeatureExtraction(covars,probVector,ratio=0.2):\n",
    "    featureList=[]\n",
    "    names=covars.columns\n",
    "    rlasso = RandomizedLasso()\n",
    "    rlasso.fit(covars,probVector)\n",
    "    sortedFeatures=sorted(zip(map(lambda x: round(x, 4), rlasso.scores_), \n",
    "                 names), reverse=True)\n",
    "    for i  in range(20):\n",
    "        f=sortedFeatures[i]\n",
    "        if f[0]>ratio:\n",
    "            featureList.append(f[1])\n",
    "    return featureList"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.Significance and effect test\n",
    "Because every mutation context can only be compared to mutations of the same class (we might want to change this to transitions/transversions to get more power) the data is devided to different mutation Types and then we preform a regression when the covars are only features we obtained in the randomized lasso test.\n",
    "\n",
    "In order to further decrease the effect of dependencies between the features I've used <a  href=\"http://statsmodels.sourceforge.net/devel/generated/statsmodels.regression.linear_model.WLS.html\">WLS method</a>  instead of OLS which puts weights on the different expainatory variables based on the covariance matrix of the variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def featureToMotiff(feature,mt,k=5,toString=True):\n",
    "    kmer=['X' for i in range(k)]\n",
    "    oldnuc=mt[0]\n",
    "    kmer[k//2]=oldnuc\n",
    "    positions=feature.split('__')[0]\n",
    "    nucs=feature.split('__')[1]\n",
    "    positions=[int(x[1]) for x in positions.split(\"_\")]\n",
    "    #print positions\n",
    "    #print nucs\n",
    "    for i in range(len(positions)):\n",
    "        kmer[positions[i]-1]=nucs[i]\n",
    "    if toString:\n",
    "        return \"\".join(kmer)\n",
    "    else:\n",
    "        return kmer\n",
    "\n",
    "def arrowBySign(float):\n",
    "    if float>0:\n",
    "        return \"\\u2191\"\n",
    "    else:\n",
    "        return \"\\u2193\"\n",
    "\n",
    "def regressOverDataFile(data,pvalCutOff=10**-4,justName=False,rawMotif=False):\n",
    "    mutationTypes=set(data['mutationType'].values)\n",
    "    #print mutationTypes\n",
    "    motifMap={}\n",
    "    for mt in mutationTypes:\n",
    "        motifMap[mt]=[]\n",
    "        #print \"-\"*20\n",
    "        #print mt\n",
    "        covars,probVector,mutData=createCovarsAndProbVectorForMutationType(data,mt)\n",
    "        features=randomLassoFeatureExtraction(covars,probVector)\n",
    "        if len(features)==0:\n",
    "            #print \"No segnificant features\"\n",
    "            continue\n",
    "        predictors=\" + \".join(features)\n",
    "        #print predictors\n",
    "        result = sm.wls(formula=\"prob~ {} +1\".format(predictors), data=mutData).fit()\n",
    "        sigFeatures=result.pvalues.sort_values()[1:5].keys()\n",
    "        #print(\"model r^2 {}\".format(result.rsquared_adj))\n",
    "        for f in sigFeatures:\n",
    "            if result.pvalues[f]<pvalCutOff:\n",
    "                if rawMotif:\n",
    "                    if justName:\n",
    "                        motifMap[mt].append(f)#,result.pvalues[f],result.params[f],result.conf_int().loc[f][0],result.conf_int().loc[f][1])\n",
    "                    else:\n",
    "                        motifMap[mt].append((f,result.pvalues[f],result.params[f],result.conf_int().loc[f][0]\\\n",
    "                                             ,result.conf_int().loc[f][1]))\n",
    "                else: \n",
    "                    if justName:\n",
    "                        motifMap[mt].append((featureToMotiff(f,mt)+arrowBySign(result.params[f]).decode('unicode-escape')))#,result.pvalues[f],result.params[f],result.conf_int().loc[f][0],result.conf_int().loc[f][1])\n",
    "                    else:\n",
    "                        motifMap[mt].append((featureToMotiff(f,mt),result.pvalues[f],result.params[f],result.conf_int().loc[f][0]\\\n",
    "                                             ,result.conf_int().loc[f][1]))\n",
    "    return motifMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "motiffMap=regressOverDataFile(data,rawMotif=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.Summarize different passeges and mutation types \n",
    "I created another function to give us the effect across diffetent passeges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def fileToPassege(fileName):\n",
    "    return fileName.split(\".\")[0][-2:]\n",
    "\n",
    "def regressOverFolder(folder,justNameVar=True,rawMotif=False):\n",
    "    os.chdir(folder)\n",
    "    files=glob.glob(\"*.csv\")\n",
    "    passeges=[fileToPassege(fileName) for fileName in files]\n",
    "    motifTable=pd.DataFrame('',index=passeges,columns=MUTATIONS)\n",
    "    #print motifTable\n",
    "    for filee in files:\n",
    "        data=pd.read_csv(filee)\n",
    "        #print filee\n",
    "        data=getDummyVarPositionForK(data,3)\n",
    "        data=cleanDataMatrix(data)\n",
    "        fileMotifMap=regressOverDataFile(data,justName=justNameVar,rawMotif=rawMotif)\n",
    "        for mut in fileMotifMap.keys():\n",
    "            if len(fileMotifMap[mut])>0:\n",
    "                motifTable.loc[fileToPassege(filee)][mut]=fileMotifMap[mut]\n",
    "    return motifTable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a summary of all significat motifs across generations, arrows imply if the motif lowers or highers the substitution rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AC</th>\n",
       "      <th>GT</th>\n",
       "      <th>AG</th>\n",
       "      <th>CA</th>\n",
       "      <th>CG</th>\n",
       "      <th>GC</th>\n",
       "      <th>AT</th>\n",
       "      <th>GA</th>\n",
       "      <th>TG</th>\n",
       "      <th>CT</th>\n",
       "      <th>TC</th>\n",
       "      <th>TA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[XTAGX↑]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[CTGXT↓]</td>\n",
       "      <td></td>\n",
       "      <td>[XXCCX↓, CXCCA↓]</td>\n",
       "      <td>[XXTAA↑]</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2</th>\n",
       "      <td>[XCAXX↑]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[CXCCA↓]</td>\n",
       "      <td></td>\n",
       "      <td>[XGGTG↓]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[XXCCA↓]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3</th>\n",
       "      <td>[XCAXX↑]</td>\n",
       "      <td></td>\n",
       "      <td>[XTAGX↑]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[XXATG↓, XTACT↓]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[XXTAA↑]</td>\n",
       "      <td>[XXTTG↓, CXTAG↓]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4</th>\n",
       "      <td>[XCAXX↑]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[XXCCA↓, XTCGX↑]</td>\n",
       "      <td>[XXTAX↑]</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P5</th>\n",
       "      <td>[XCAXX↑]</td>\n",
       "      <td></td>\n",
       "      <td>[CAATX↓, XTAGX↑]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[XXTCC↑]</td>\n",
       "      <td>[CXCCA↓]</td>\n",
       "      <td>[XXTAA↑]</td>\n",
       "      <td>[CXTAC↓]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P6</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[CTGXT↓]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[XXCCX↓]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7</th>\n",
       "      <td>[AXATA↑]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[GCCXX↑]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[XCGXX↑]</td>\n",
       "      <td></td>\n",
       "      <td>[XXCCX↓, CXCCA↓]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          AC GT                AG        CA        CG        GC  \\\n",
       "P1                       [XTAGX↑]                                 \n",
       "P2  [XCAXX↑]                       [CXCCA↓]            [XGGTG↓]   \n",
       "P3  [XCAXX↑]             [XTAGX↑]                                 \n",
       "P4  [XCAXX↑]                                                      \n",
       "P5  [XCAXX↑]     [CAATX↓, XTAGX↑]                                 \n",
       "P6                                                     [CTGXT↓]   \n",
       "P7  [AXATA↑]                                 [GCCXX↑]             \n",
       "\n",
       "                  AT        GA        TG                CT        TC  \\\n",
       "P1                    [CTGXT↓]            [XXCCX↓, CXCCA↓]  [XXTAA↑]   \n",
       "P2                                                [XXCCA↓]             \n",
       "P3  [XXATG↓, XTACT↓]                                        [XXTAA↑]   \n",
       "P4                                        [XXCCA↓, XTCGX↑]  [XXTAX↑]   \n",
       "P5                              [XXTCC↑]          [CXCCA↓]  [XXTAA↑]   \n",
       "P6                                                [XXCCX↓]             \n",
       "P7                    [XCGXX↑]            [XXCCX↓, CXCCA↓]             \n",
       "\n",
       "                  TA  \n",
       "P1                    \n",
       "P2                    \n",
       "P3  [XXTTG↓, CXTAG↓]  \n",
       "P4                    \n",
       "P5          [CXTAC↓]  \n",
       "P6                    \n",
       "P7                    "
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regressOverFolder(r'C:\\Users\\Guyling1\\ContextProject',rawMotif=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "summary=regressOverFolder(r'C:\\Users\\Guyling1\\ContextProject',justNameVar=False)\n",
    "rawMotifSummary=regressOverFolder(r'C:\\Users\\Guyling1\\ContextProject',justNameVar=True,rawMotif=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see more details about the different motifs ($P-value$,$\\beta$ coeff,$CI_\\alpha[low]$,$CI_\\alpha[high]$)) for $\\alpha=0.05$\n",
    "\n",
    "Just change the Passege and mutation Type to see the motiff you want"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('XXTAX',\n",
       "  6.50253875486119e-08,\n",
       "  0.6057082529138077,\n",
       "  0.39303962494082784,\n",
       "  0.81837688088678751)]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary.loc['P2']['TC']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.Motif Analsys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparing Histograms of sequences with the motif and sequences without the motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def seperateMotifAndNonMotif(motif,data,mutationType):\n",
    "    data=data[data['mutationType']==mutationType]\n",
    "    nonMotifData=data[data[motif]==0]\n",
    "    motifData=data[data[motif]==1]\n",
    "    return motifData,nonMotifData,motif\n",
    "\n",
    "def histOfMotfAndNonMotif(motif,nonMotif,motifName):\n",
    "    motifProb=motif.prob\n",
    "    nonMotifProb=nonMotif.prob\n",
    "    #print nonMotifProb\n",
    "    df=pd.DataFrame({'Motif':motifProb,'nonMotif':nonMotifProb},columns=['Motif','nonMotif'])\n",
    "    plt.figure()\n",
    "    df.plot.hist(alpha=0.5,bins=50,normed=True)\n",
    "    plt.title(\"{} motif analysis\".format(motifName))\n",
    "    plt.text(-11,1.0,\"motif mean=%.2f\" % motifProb.mean())\n",
    "    plt.text (-11,0.8,\"nonMotif mean=%.2f\" % nonMotifProb.mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 388,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.figure.Figure at 0x12b27eb8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY4AAAEPCAYAAABV6CMBAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XlcVOX+B/DPmYFhRIZVTAQXFHFBwwVzwesWasutNJXq\nes01l3LrpuZWklv6S0pTseRS2jUz1NQsyy0zxSU0ScWF8IYLioqIgALDzDy/P7yeHGeQOQozg37e\nrxevF+c5z3nO95xZvnOec85zJCGEABERkY1Ujg6AiIgqFyYOIiJShImDiIgUYeIgIiJFmDiIiEgR\nJg4iIlKEiYPITt577z2Ehoaala1ZswYhISFwdXXF4MGDHRTZX1QqFVatWlVu7XXp0gXDhg0rt/bI\nOTBx0H0bNGgQVCoVVCoVXF1dUbduXYwcORI5OTlW6w8cOBAqlQpz5swptxhMJhPi4uLQvn17eHt7\nw8PDA2FhYRgyZAgOHTpUbutRIikpCSqVCmfPnjUrnzBhAvbv3y9Pm0wmDBkyBC+//DLOnTuHhQsX\n2jvUCrd+/Xp8+OGHjg6DyhkTBz2Qjh074tKlSzhz5gwWLVqEb775BgMGDLCot3z5cqSmpiIwMLDc\n1m0wGPDMM89gypQpePHFF7F161acPHkS//nPf1CvXj289dZb5bYuJYQQkCTJotzd3R2+vr7y9IUL\nF1BQUICnn34aNWrUgE6ns2eYdnE7mdNDRhDdp4EDB4pu3bqZlc2ePVu4uLiIoqIiuSw1NVUEBASI\n9PR0UbduXTF79uxyWf/8+fOFWq0WBw8evK/l69atK9555x0xcuRI4e3tLapXry6WLFkiiouLxejR\no4WPj48IDAwUixcvNlvu4sWL4qWXXhLe3t6iSpUqonPnznIMGRkZQpIkoVKphCRJQpIk0aVLFyGE\nENOnTxchISFCCCGWL19uVk+lUoldu3ZZjXPbtm2ic+fOwtfXV3h5eYlOnTqJX3/91ayOJEkiLi5O\n9O/fX+h0OhEUFCTef/99szqrVq0Sbdq0EV5eXqJatWri2WefFWlpaRbtfPnll0KIW69v9+7dLeLp\n0qWLGDp0qBBCiPPnz4vevXuLatWqCa1WK+rXry/mz58v1+3cubN47bXX5Ondu3eLyMhIodPphE6n\nE82bNxdbt24t5RUiZ8UjDipXWq0WJpMJBoMBAFBYWIiXXnoJ8+fPR/369ct1XStXrkRUVBRatWp1\n320sXrwYoaGhOHToEMaOHYtRo0ahV69eqFevHg4ePIhRo0ZhzJgxOHnypLzMCy+8gLS0NGzevBnJ\nycl47LHH0K1bN+Tk5KBWrVrYuHEjAODgwYPIysrCN998AwCQJEk+Enn55Zfx66+/QgiBTZs24eLF\ni2jfvr3VGAsKCvDGG2/gwIED2LdvH0JDQ/HUU0/h2rVrZvVmzJiBTp064ffff8fkyZMxZcoU7Ny5\nU56v1+vxzjvvICUlBdu3b4eLiwueffZZ+bW62/Dhw7Fjxw6cOXNGLktPT8euXbswfPhwAMDIkSOR\nl5eHn376CadOnUJCQgKCgoKstmc0GvHCCy+gXbt2SElJweHDhxETEwN3d/d7vkbkhByduajyuvuI\nIzU1VdSvX1+0b99eLhs0aJAYOHCgPF2eRxzu7u5i3LhxZmVvv/228PDwkP/OnTtX6vJ169YVvXr1\nkqdNJpPw9PQUzz//vFmZj4+PWLJkiRBCiO3btwuVSiVOnjwp1ykuLhYBAQFi5syZQggh9uzZI1Qq\nlThz5ozZ+mJiYkSDBg3k6dtHJ0lJSYq222g0Ch8fH7Fq1Sq5TJIki33RuHFjMWXKlFLbuXr1qpAk\nSezdu9esndtHHEII8fjjj4t33nlHnp40aZJo3ry5PB0eHi7ee++9Utdx5xHHtWvX7nlkRZUHjzjo\ngezcuRM6nQ7u7u54/PHHERISgpUrVwIAvvzySyQlJWHJkiUVtn5x1xidEydOxO+//46EhATcvHkT\nJpPpnsuHh4fL/0uSBH9/fzz++ONmZdWrV8fly5cBAMePH4efnx8aNmwo19FoNGjTpg1SU1PLY5Ms\nZGRkoH///mjQoAG8vLzg5eWFvLw8syOBu7cFAGrWrIlLly7J0ykpKXjxxRdRr149eHp6ok6dOpAk\nyaKdOw0fPhyff/45hBAwGo1YsWKF2VVS48aNw+zZs9G2bVtMmjQJu3fvLrUtb29vDBkyBN27d8cz\nzzyDefPmIS0tTenuICfAxEEPpG3btjhy5AhOnjyJoqIi/PjjjwgODgYAbN++HadPn4aXlxdcXV3h\n6uqKs2fP4t133y2X7onQ0FCcOHHCrMzX1xf16tWz+SS8q6ur2bQkSVbLykpAFenZZ5/F+fPnERcX\nhwMHDuD333+Hv78/9Hq9WT2NRmM2fWfchYWF6NGjB1QqFZYvX47k5GQcPHgQACzauVP//v1x/fp1\nfP/999i0aRPy8vLQr18/ef7AgQNx9uxZjBw5EllZWXj66afx6quvltresmXL8Ntvv6F79+7YtWsX\nmjZtivj4eMX7hByLiaMS2bVrF/bt2ydPf/rpp/Kv+1OnTqFFixZo1aoV/vzzT7vFVKVKFQQHB6N2\n7dpwcXExmzdnzhwcOXIEv//+u/wXEBCAUaNG4bfffpPr/f7772jXrh1atGiBJ554Qv5Cu1NaWhpa\ntGiBli1bokWLFvDy8kKdOnWwY8cOrFy50mJ5IYTF0Uh5CAsLw9WrV83OeRQXF+PAgQNo1qwZgL++\nwI1G4wOvLycnBydOnMCkSZPQrVs3NGrUCBqNRj4CstWJEyeQnZ2N2bNno2PHjmjYsCGuXr1a5j7S\n6XR4+eWXsWzZMvz73/9G37594enpaVbnsccew4ABA7B8+XIkJCTgyy+/REFBQaltNmnSBOPGjcPm\nzZsxZMgQLFu2TNG2kOO5lF2FnMXPP/8MDw8PtGvXDgDkE5QAsGHDBvTt2xdTpkxxVHgWAgICEBAQ\nYFbm6uqK6tWro1GjRnLZxIkT8d5776F79+744YcfMGHCBLOTusCto4vDhw8DuHX/Q1BQEBYsWIDC\nwkIMGjQIr776KoYPH44TJ05g2LBhaNasGSRJglqtLtdt6tq1K1q3bo1//OMfWLx4MTw9PTFz5kwU\nFxdjxIgRAIA6depApVJh8+bNiI6Ohpubm8WXra18fHzg7++P+Ph41KtXD9nZ2Xj77bcVH7HVqVMH\nbm5u+Pjjj/HWW2/hzz//xOTJk6FSlf3bcdiwYWjXrh0kScKuXbvM5o0ePRrPPPMMGjZsiMLCQqxb\ntw61a9e2egnu6dOnER8fj+eeew61atVCZmYmdu/ejYiICEXbQo7HIw47OnPmDBo3boxBgwahYcOG\n+Oc//4kdO3agQ4cOaNiwofxL+9q1a+jVqxfCw8PRvn17HDt2DGfOnMEnn3yCBQsWoGXLlkhKSsJ7\n772H2NhY/PDDD1iwYAGWLl2KJ5980mK9Op0OEydORNOmTdG9e3ckJyejS5cuCAkJwXfffQfg1pfx\nxIkT0aZNGzRv3lzuPrhx4waioqIQERGB8PBwfPvtt/K2rF+/HsePH0fTpk3x1FNPobi4uMx9YO3+\nBpVKhevXrwMAcnNzy+xm2r59O+rXr4+6devihx9+QIMGDbB7925069YNr732GtLS0iCEwJ49e+7Z\nlrVYbCnbuHEjGjVqhL///e9o06YNLl++jO3bt8v3aFSvXh3vv/8+5s6di5o1a6Jnz56KYrh7/tq1\na3H69GmEh4dj8ODBePPNNy0Sclnt+Pn5YeXKldi+fTuaNm2KiRMnIjY21iJxWGsnIiICzZo1Q8OG\nDeUfLbcJIfDmm2+iWbNm6Ny5MwoLC7F582ar7VWtWhV//PEHXnnlFTRs2BB9+/ZFhw4dsGjRonvG\nTk7IXmfh4+LixNChQ8Vbb71Vap2EhAQxevRoMX78ePHnn3/aKzS7ycjIEK6uriI1NVUIIUSrVq3E\nkCFDhBBCbNy4Ub7CZ/To0WLGjBlCCCF++ukn+SqWmJgYERsbK7d35/Td8+4kSZLYsmWLEEKIXr16\niR49egij0Sh+//13ue1ly5bJVzsVFxeLiIgIkZGRIYxGo8jPzxdCCJGdnS3fh3B7W44cOSKEECI6\nOlq+GueDDz4QLVq0sPgbO3as1fhOnDghateuLWrVqiWCgoLE2bNn77kfBw8eLF/ldD/LkzIlJSUi\nMDBQLFq0yNGhkJOw2xFHly5dMHXq1FLnHz58GJcuXcLHH3+MYcOGKTphVlFXs5S3tLQ0BAcHo0mT\nJgBu9ZffPkJo1qwZMjIyAAB79uxB//79Adzabzk5OffsMy6Lm5sbunfvLq+nU6dOUKlUaNasmXxF\nzdatW/HFF1+gRYsWCA8PR05ODv744w+YTCZMmjQJ4eHhiIqKwoULF+T+9eDgYLlfv1WrVnL848eP\nx2+//Wbxt2DBAqvxLV26FAsXLsTZs2fx0Ucf3XPMppKSEnz77bfo27ev/LorWd4RKsP701qMQghc\nvnwZ77//Pm7evImBAwfaP7C7VIZ9CTz8cdrtHEejRo1w5cqVUucnJyejU6dOAIAGDRrg5s2byM3N\nhbe3d5ltp6amIiwsrNxirShpaWlwc3OTp1UqlTytUqlKvRHrQd15ldCd65QkSV6nEAKLFi1Ct27d\nkJiYiOjoaADAihUrcPXqVRw+fBgqlQrBwcEoKioCALNtUavVcvn8+fPx5ZdfWsTRqVMnLFiwAIMH\nD5bPV6Snp6OgoACfffaZXK+goEA+N3F3X77BYEBJSQm2b98OtVqNsLAwrFixQh7nqU+fPhgyZMgD\n7K3yVxnen9ZiPHv2LIKDg1GzZk18/vnnTjF0SGXYl8DDH6fTnBzPycmBn5+fPO3r64ucnBybEkdl\nImy40udvf/sbVq5ciWnTpuHnn39GtWrV4OHhAZ1Oh7y8vHJd5+15PXr0QFxcHLp06QIA+OOPPxAY\nGIjr16+jevXqUKlU2Llzp9k1/6W1O378eIwfP77Udd5OEjdu3MClS5fQo0cPzJgxA23atEFSUhI+\n+OADbNiwweqyY8eORceOHfH888/j+++/BwAEBgZi165d6NSpE3bs2GExAi3dnzp16jj0MmRyXk6T\nOB4Vd54sLO2EZkxMDAYPHozw8HBUrVoVK1asAAA899xz6NOnD7799lssWrSozBOiZa3nznlDhw5F\nRkYGWrZsidzcXDRo0AAbNmxAv3798NxzzyE8PBwRERFo3LixTe3aomrVqqhXrx5WrFiBMWPGwGg0\nQqvVYsWKFahXrx4uXryI1157TT6Bf/PmTezfvx+rVq1C1apV5Xbi4+PNluflnUQVSxK2/AQuJ1eu\nXMG8efMwf/58i3nLli1D06ZN5fF6xo0bh5iYGKtHHKmpqWZ9c7e7VYiISJnExET5/7CwMJu6rux6\nxCHucVNWREQEtmzZgvbt2yMtLQ1Vq1YttZvK2sZduHCh3OMtbzqdDvn5+Y4Oo0yMs3xVhjgrQ4wA\n4yxvNWvWvK8f3nZLHAsXLsTx48eRn5+PkSNHIjo6GgaDAZIkISoqCi1btsThw4cxevRoaLVajBw5\n0l6hERGRAnZLHGPHji2zjrNdDUMPHykvF8jPtZyh84bwfLguxCCqKDw5To+W/FwYD+21KFa3ag8w\ncRDZhEOOEBGRIjziIKIK5+Hh8cCXbwO3bjStDM9md7Y4hRAPNPrE3Zg4iKjCSZJUKa4yeliVdxJj\nVxURESnCxEFERIowcRARkSJMHEREFWDSpEnyqM3ArZGmmzdvjoYNGyI318q9RJUIT44TkUOUejNm\neVFwU2ebNm1w5coVHDp0CD4+PnJ59+7dcfz4cRw4cOCeT5NMTEzEV199hfXr18tlc+fOlf83GAyY\nMWMGvv/+e7PHJldWTBxE5Bil3IxZXpTc1ClJEmrVqoWNGzfKD6w6efIkioqKbLqMWAhxz3qXL1+G\nXq9HgwYNbIrH2bGriogIQO/evbFmzRp5es2aNejbt688nZ+fjzFjxuDxxx9H27Zt5W6o9PR0TJky\nBYcOHUJoaKg8AOubb76JDz74AP/973/lh9Q1adIEL730kh23qmIwcRARAWjZsiUKCgqQnp4Ok8mE\nb7/9Fi+++KI8f+rUqbhx4wYOHDiAtWvXYu3atfj6668REhKC999/H61atUJaWprF41jr1auHnTt3\nArh1FPP111/bdbsqAhMHEdH/3D7q+OWXX9CgQQPUqFEDwK1zFJs2bcLkyZNRpUoVBAUFYfjw4Vi7\ndq2i9u34+KMKxXMcRET/07t3b7z44os4d+4c+vTpI5fn5OTAYDCYnSAPCgpCVlaWI8J0OB5xEBH9\nT2BgIGrVqoWdO3fi6aeflsv9/Pzg6uqKzMxMuez8+fPyEUl5jMNVmTBxEBHd4cMPP0RiYiKqVKkC\n4Fb3klqtxrPPPot58+bhxo0bOH/+POLj4+WjEn9/f1y8eBElJSWltvuwdFMB7KoiIkfRed+6ZLYC\n27fVnUcMtWvXtjpv1qxZmDZtGtq1awetVot+/frJV0hFRkYiNDQUzZs3h1qtxpEjR+65jspOEg9J\nGuQzx8vPwxynlJlR6oOcRGDdcorMXGXYnxUdY2XYBw+z0vZ/zZo176s9dlUREZEiTBxERKQIEwcR\nESnCxEFERIowcRARkSJMHEREpAgTBxERKcLEQUREijBxEBFVIpmZmWjYsKE8hEl2djZefPFFNGrU\nCDNnzrRLDBxyhIgcIrfIiNxiY4W17+2mhrdWXWHtlyUxMRH/+te/8Nprr2H69Oly+ZYtWzBkyBBE\nR0fjww8/LLOdtm3bYv78+ejQoQOAWwMxnjp1Sp6/cuVK+Pn54eTJk+W/EaVg4iAih8gtNmLfmesV\n1n67Ol4OTRwAUKdOHWzatAnvvPMOVKpbHTxr165F/fr1y20dmZmZCA0NLbf2bMGuKiJ65LVt2xaf\nfPIJoqKi0KRJE7z++uvQ6/UAgC+//BKRkZFo2rQpBg8ejEuXLsnLBQUF4T//+Q86dOiAsLAwTJ06\n1azd6tWro3Hjxvj5558BALm5uTh48CC6d+9uVm/r1q3o2rUrwsLC0LdvX6SnpwMAxowZg8zMTAwc\nOBANGzbEJ598gvPnzyMoKAgmkwlvvvkm1qxZg7i4ODRs2BB79uypwL30FyYOIiIA3333Hb766ivs\n27cPx48fR2JiIpKSkjB37lwsW7YMhw8fRmBgIF5//XWz5Xbs2IEff/wRW7duxaZNm7Br1y55niRJ\n6NOnj/ws840bN6JHjx5wdXWV65w+fRpvvPEGZsyYgSNHjqBr164YMGAADAYDPv74YwQGBmLFihU4\ndeoURowYIbcLAB999BF69eqF119/HadOnZK7syoaEwcREYChQ4fC398fXl5e6NatG44dO4b169fj\nlVdeQVhYGFxdXTF58mQcOnTI7IFOo0aNgoeHBwIDA9G+fXuLZ4736NED+/fvR35+PtauXWv2ZEEA\n2LRpE6KiotChQweo1WqMGDECRUVFOHjwoFzH2QYxZ+IgIgJQrVo1+f8qVargxo0buHTpktnjYt3d\n3eHj44OLFy/KZf7+/hbL3Umr1eLJJ5/EwoULkZubi4iICLP5ly5dQlBQkDwtSRJq1qzp1I+lZeKo\npPbt24egoCCsXr1aLktNTUVQUBA+/fTTey6bmJiIy5cvy9MTJ06U+1S/++47tG7dGtHR0RUTeAXa\nsGEDoqKi0K1bN/Tv3x/Xrl2zqGMwGPDmitXoNisWT86YjyVbfpLn9enTBx07dkT37t3Ro0cP5OTk\n2DN8cjKSJKFGjRo4f/68XHbz5k1cu3YNAQEBitrq3bs3li1bht69e1vMe+yxx8zWAdx6vtDtdTjj\nA6CYOCqxRo0aYdOmTfL0xo0bERYWVuZyiYmJZr9m/u///g8hISEAgNWrV2Px4sVITEws/4ArkNFo\nxPTp07F27Vps27YNjRo1wvLlyy3qbdqxEyUGI7ZNewvfTRqLL3fvR2bOXwkmLi4OW7duxZYtW+Dr\n62vHLSBn9MILLyAxMRHHjx9HcXEx5s6di5YtW5odhdiiXbt2+OqrrzBo0CCLec899xx27NiBpKQk\nGAwGfPLJJ9BqtWjVqhWAW0c0Z8+eNVvG0V1XvBzXjs6ePYtevXrhiSeewMGDBxEQEIDPPvsMbm5u\nSE1NxaRJk1BUVIS6desiNjYWnp6e6NOnD1q2bIm9e/ciLy8PsbGxaN26NYBb13PfuHEDV69ehZ+f\nH3bu3Iknn3xSXt+xY8cwefJkuc358+djz549OHLkCEaPHg2tVouNGzfin//8J6ZPn47t27fj119/\nxRtvvIGoqCizK0T27dsnx3Tq1Cn8/e9/R6NGjZCQkICioiJ89tlnqF27NnJycvD222/Lh/IxMTGI\niIhASkoK3n33Xej1emi1Wnz44YeoV68eEhMTsW3bNhQWFuLMmTN46qmnLK5MscXtD9KNGzfg5eWF\n/Px8BAcHW9STJAk39XoYTSYU6fXQuLjAQ6uV55tMJsXrpvvj7aZGuzpeFdq+rUr7Vd+hQwdMmDAB\nr732Gq5fv46IiAjExcWVuZw1kZGRVsvr16+PRYsWYdq0abh06RKaNGmC5cuXw8Xl1tfzqFGjMG3a\nNMyePRtjxozBs88+6/CjELs+OjYlJQXLly+HEAJdunRBz549zebfvHkTixYtQnZ2NkwmE5577jl0\n7tzZprYrw6Njr127hhYtWuCHH35A48aNMWLECPTo0QO9evVCVFQU5syZgyeeeALz589HQUEBYmJi\n0KdPH4SHh+Odd97BTz/9hGXLlmH16tXYt28fPvnkE3Tu3BmSJCEsLAyrV69GUFAQ3N3dMXz48Hu2\nGRMTg6ZNmwK41UUzffp0NGvWDH369MH//d//oV69emax79u3D0OHDsWuXbvg6emJdu3aoV+/fvjX\nv/6FhIQEnDt3DjExMRg1ahQGDBiA1q1bIzMzE/369cPPP/+MGzduoEqVKlCpVNi9eze++OILxMfH\nIzExEQsXLsTWrVvh4uKCjh07YsOGDQgICMDIkSPx3//+12I/Dhs2DL1797Z4HOb333+Pt956C1Wr\nVkVwcDDWrFlj8QEznknHqDFjkXTyDxSVGDC9z3N4pUMbqFu1R++x43Ht2jW4uLjg6aefxrhx48rl\nda8Mj03lo2MfbuX96Fi7HXGYTCYkJCTg3XffhY+PDyZPnozWrVubHfJt2bIFtWrVwttvv428vDyM\nGzcOf/vb36BWO/YmnvJUq1YtNG7cGADw+OOP49y5c8jPz0d+fj6eeOIJAEDfvn3ly+4A4Omnn5br\n39kXKkkSnnvuOYwYMQLp6el44YUX5CsxymrzXr8XSpsXHh4un0CsW7cuOnbsCOBWl9m+ffsAALt3\n78Yff/xhdgRQWFiI69evY+zYsfjzzz8hSRIMBoPcbmRkJKpWrQoACA0Nxfnz5xEQEIClS5fee2fe\nwWAw4IsvvsC2bdtQq1YtTJs2DR9//DHGjh1rVu/w8RNwUanw27x3ce3GTfSOjUOHxg1QF8CSJUvw\n2GOP4ebNmxg6dCjWrVtntU+a6FFnt8SRnp6OgIAA+QqEyMhIJCcnmyUOSZJQWFgIACgqKoJOp3uo\nkgYAuLm5yf+rVCoYjbeGXLjXF/ntZe6sf1u1atXg4uKC3bt3Y+bMmRV6CZ9Go5H/lyTJLK7biUAI\nge+++87sOnUAmDJlCiIjI/Hvf/8b58+fR9++fS2273Zbt7dx5MiROH36tFk7kiRh2LBh6NWrFzp0\n6ACTyYTu3bsjKioKwK3EDNzqN16yZInFNqzfshWdwxpCpVLBT+eBiPp1ceTMedTFrZOUwK0rZ3r2\n7ImUlBQmDiIr7JY4cnJy4OfnJ0/7+vrKV/Lc9tRTT2HevHkYPnw4ioqKyq2rwJlY+zLX6XTw9vZG\ncnIyWrdujXXr1qFt27Y2Lz9hwgRcvXrVrFvmXm16eHhUWLdBp06dkJCQIB/dpKamIiwsDAUFBahR\nowYA4Ouvv7aprbKOOPbs2SNvx6VLl5Ceno6cnBz4+vril19+QYMGDSyWCXysBpJ+S0avJ1riZrEe\nh/88i6FdO8JoNCL3f8uWlJRg+/bt8hEVEZlzqpPjKSkpCA4OxvTp05GVlYVZs2Zh/vz50N5x8hK4\n9WV050020dHR0Ol09g5XsYKCAqjVajlWrVYLo9EInU6HZcuWYdy4cfKJ7Li4OOh0Ori6usLd3R06\nnQ56vV5e3t3dHS4uLtDpdGbngTQaDbRa7T3bHDBgAKZMmQJ3d3ds27bNbB2urq5wdXW12J93rg+A\n2TJ3zouNjcVbb72FHj16wGg0on379vjoo4/wr3/9CyNGjMDixYvRvXt3qFQq6HQ6aLVas/Xd2W5Z\nNBqNXE+n02Hy5Mno06cPNBoNatWqhaVLl0Kn0+GHH37A4cOHMWXKFAzv9w+M2r8XUTNvDS7X72/t\n8HhwHeihwquvvgqDwQCj0YjOnTtjxIgR5XIS8s44nVVFx/iw9RxUNnd+79ztzisow8LCbLoy024n\nx9PS0rBmzRr5ipkNGzYAgNkJ8rlz56Jnz55o1KgRAGDGjBno16+fTQOCVYaT45XlBOHDHKeUmQHj\nob0W5epW7SEC65ZTZOYqw/7kyfGHW3mfHLfbfRwhISHIysrClStXYDAYkJSUZHEHZbVq1XD06FEA\ntwYDu3jxotzvTEREzsFuXVUqlQpDhgzBrFmzIIRA165dERQUhG3btkGSJERFRaF3796Ii4vD+PHj\nAQD9+vWDh4eHvUIkogoihCiXrjC1Wm1xgYgzcrY4y7tjya73cVQkdlWVn4c5TnZVWVcZYgQYZ3lz\n+q4qIiJ6ODBxEBGRIkwcRESkCBMHEREpwsRBRESKMHEQEZEiTBxERKQIEwcRESnCxEFERIo41ei4\nRPci5eUC+bnydIlGC0lfBOi8ITy97bJOABW6PqLKgImDKo/8XLPhQgwaNxj1xVC3ag9U1Bf5XesE\nULHrI6oE2FVFRESKMHEQEZEiTBxERKQIEwcRESnCxEFERIowcRARkSJMHEREpAgTBxERKcLEQURE\nijBxEBE9+psuAAAarUlEQVSRIkwcRESkCBMHEREpwsRBRESKMHEQEZEiHFadqJzkFhmRW2y0KA+Q\niuHmgHiIKgoTB1E5yS02Yt+Z6xblXTRuqFHFAQERVRB2VRERkSJMHEREpAgTBxERKcLEQUREijBx\nEBGRIkwcRESkCBMHEREpYnPi2Lx5M/Ly8ioyFiIiqgRsvgHw2LFj+OqrrxAWFoaOHTuidevWcHV1\nVbSylJQULF++HEIIdOnSBT179rSok5qaihUrVsBoNMLT0xPTp09XtA4iIqpYNieOiRMnIj8/H0lJ\nSfj+++8RHx+PNm3aoGPHjmjSpEmZy5tMJiQkJODdd9+Fj48PJk+ejNatWyMwMFCuc/PmTSQkJGDa\ntGnw9fXlEc5DRMrLBfJzzQt13hCe3g/edkkJkJlRIW0TkSVFQ47odDo89dRTeOqpp3DmzBksXrwY\nO3fuRLVq1fDkk0/imWeegVartbpseno6AgIC4O/vDwCIjIxEcnKyWeLYs2cP2rRpA19fXwCAp6fn\n/W4XOZv8XBgP7TUrUrdqD5TDl7u4kQ/j6ZMV0jYRWVI8VtXRo0exe/duJCcno379+hg1ahSqVauG\nzZs3Y86cOZgxY4bV5XJycuDn5ydP+/r6Ij093azOhQsXYDQa8d5776GoqAhPP/00OnbsqDREIiKq\nQDYnji+++AJ79+6Fu7s7OnbsiNjYWPnIAAAaNGiAQYMGPVAwJpMJf/75J959910UFxdj2rRpCA0N\nRY0aNR6oXSIiKj82J46SkhKMHz8eISEh1htyccHcuXNLXd7X1xfZ2dnydE5OjlniuV1Hp9NBo9FA\no9GgcePGyMjIsEgcqampSE1Nlaejo6Oh0+ls3RSH0Wg0j2ycJRotDBrzwcVdNFq4KljP3W2o1Wpo\nNG5QubhCbWPb1uJQWr+0utrCAmistK12cYFOZ70L11k8yu/NilBZ4gSAxMRE+f+wsDCEhYWVuYzN\niaNXr17QaDRmZQUFBdDr9XICuPN8xd1CQkKQlZWFK1euwMfHB0lJSRg7dqxZndatW+Ozzz6DyWRC\nSUkJ/vjjD/z973+3aMvaxuXn59u6KQ6j0+ke2TglfRGM+mKzMqO+CEUK1nN3GxqNG/T6YqgNJTa3\nbS0OpfVLq1uk10NvrW2Dwelf90f5vVkRKlOc0dHRipezOXF88MEHGDlyJDw8POSynJwcfPLJJ5gz\nZ06Zy6tUKgwZMgSzZs2CEAJdu3ZFUFAQtm3bBkmSEBUVhcDAQISHh2P8+PFQqVSIiopCUFCQ4o0i\nIqKKY3PiuHDhAmrXrm1WVrt2bWRmZtq8subNm2PhwoVmZd26dTObfv755/H888/b3CYREdmXzXeO\ne3p6Iisry6wsKyur0vTjERFR+bD5iKNLly6IjY3Fyy+/jMceewxZWVn4+uuv0bVr14qMj4iInIzN\niaNnz55wcXHBf/7zH1y9ehV+fn7o2rWr1ZPXRET08LI5cahUKp5/ICIiZXeOX7hwARkZGSgqKjIr\nZ3cVEdGjw+bE8c0332DdunWoU6cO3NzMb3Ji4iAienTYnDhuj0VVp06dioyHiIicnM2X42o0mnve\nGU5ERI8GmxPHSy+9hM8++wzXrl2DyWQy+yMiokeHzV1VcXFxAIAdO3ZYzPv666/LLyIiInJqNieO\nxYsXV2QcRERUSdicOG4/uc9kMuH69evw8fGpsKCIiMh52Zw4bty4gX//+9/Yv3+/fAf5wYMHkZ6e\njpdffrkiYyQiIidi88nx+Ph4uLu7Iy4uDi4ut/JNaGgo9u7dW8aSRET0MLH5iOPo0aP49NNP5aQB\n3Box9/r16xUSGBEROSebjzjc3d0tnmiVnZ3Ncx1ERI8YmxPHk08+idjYWBw7dgxCCKSlpWHJkiUW\nD2IiIqKHm81dVS+88AI0Gg0SEhJgNBqxdOlSREVF4ZlnnqnI+IiIyMnYnDgkScIzzzzDREFE9Iiz\nOXEcO3as1HlNmzYtl2CIiMj52Zw4li5dajadl5cHg8EAPz8/3lVOTkcqKQEyMyxnFBbaPRaih43N\niWPJkiVm0yaTCevWrUOVKlXKPSiiByVu5MN4+qRFubp+IwdEQ/RwsfmqKosFVSq8+OKL2LhxY3nG\nQ0RETu6+EwcAHDlyBCrVAzVBRESVjM1dVSNHjjSb1uv10Ov1GDp0aLkHRUREzsvmxDF69GizaTc3\nNwQEBMDd3b3cgyIiIudlc+Jo0qRJRcZBRESVhM2JY9GiRZAkqcx6o0aNeqCAiMpDrsYTOb7BFuW+\nGk/oHBAP0cPE5sRRtWpV7Nq1C61atUK1atWQnZ2NQ4cOoVOnTtDp+FEk55JrVGHfxSKL8shqKiYO\nogdkc+K4ePEiJk2ahMaNG8tlJ0+exLp16zB48OAKCY6IiJyPzdfSpqWloUGDBmZlISEhSEtLK/eg\niIjIedmcOIKDg/HVV19Br9cDuHU57urVq1G3bt2Kio2IiJyQzV1Vr7/+Oj7++GMMGDAAHh4eKCgo\nQP369TFmzJiKjI+IiJyMzYmjevXqmDVrFrKzs3Ht2jX4+PigWrVqFRkbERE5IUXjheTn5+P48eM4\nfvw4qlWrhpycHFy9erWiYiMiIidkc+I4fvw4xo0bh927d2PdunUAgKysLMTHx1dYcERE5HxsThzL\nly/HuHHjMHXqVKjVagC3rqo6ffq0zStLSUnBuHHjMHbsWGzYsKHUeunp6XjllVdw4MABm9smIiL7\nsDlxXLlyBc2aNTMrc3FxgdFotGl5k8mEhIQETJ06FbGxsUhKSkJmZqbVeqtWrUJ4eLitoRERkR3Z\nnDiCgoKQkpJiVnb06FHUrl3bpuXT09MREBAAf39/uLi4IDIyEsnJyRb1fvzxR7Rt2xaenp62hkZE\nRHZk81VV/fv3x7x589CiRQvo9XosW7YMhw4dwoQJE2xaPicnB35+fvK0r68v0tPTLeokJydj+vTp\nFvPIMXKLjMgttjyq9HZTw1urtr0dSWsxdpSvpIXXA0dofVyqYhc3q3X1mio4Y20Mq3KKhehRYHPi\nCA0NxQcffIDdu3dDq9WiWrVqmDNnjlkyeFDLly9Hv3795GkhRLm1Tfcnt9iIfWeuW5S3q+OlLHGU\nwGLsqMhaKJ/EYWVcqsbVrR9MFxiAE9bGsCqnWIgeBTYlDpPJhBkzZmDq1Kl44YUX7mtFvr6+yM7O\nlqdzcnLg6+trVue///0vFixYACEE8vPzcfjwYbi4uCAiIsKsXmpqKlJTU+Xp6OjoSjHQokajqXRx\nagsLoNFY/nrXatyg03nY3KaLS7Z8UcVfZS6K9keJRgvDHbGo1WpoNG5Qq9QWbUuSZFF2r/LSYrl7\nnQDgotHC1Urd0vaV2sUFOp229A1zApXxvenMKkucAJCYmCj/HxYWhrCwsDKXsSlxqFQqXL58+YGO\nAEJCQpCVlYUrV67Ax8cHSUlJGDt2rFmdxYsXy//HxcWhVatWFkkDsL5x+fn59x2bveh0ukoXZ5Fe\nD72+2KJOkb4Y+fm2vx8MBoPFhRQGg0HR/pD0RTDeEYtG4wa9vhhGk9GibSGE1Qs3SisvLZa71wkA\nRn0RiqzULW1fGRVupyNUxvemM6tMcUZHRytezuauqj59+iA+Ph7R0dEW3VO2PHdcpVJhyJAhmDVr\nFoQQ6Nq1K4KCgrBt2zZIkoSoqCjFwRMRkf3ZnDg+/fRTAMAvv/xiMe/rr7+2qY3mzZtj4cKFZmXd\nunWzWvf111+3NTQiIrKjMhNHbm4uvL29zbqRiIjo0VVmH9Pt8xD+/v7w9/fHihUr5P9v/xER0aOj\nzCOOu0+I33k1Ez38pMJCIC/XSrkG8NI4IKKKIZmMkDIzLGcUFlrWLSkBrNSV1D7lHxiREyozcUiS\nZI84yFkVF0JcPG9ZXkuHh+nOB1FcDOPR/Rbl6vqNLOveyIfx9EnLRsK7ArD93haiyqrMxGE0GnHs\n2DF52mQymU0DQNOmTcs/MiIickplJg4vLy8sXbpUnvbw8DCbliSJJ86pXJU2zImPpIUz3FJlbYgT\nACiWXABwtAN6+JWZOJYsWWKPOIhkpQ1z0t4LzpE4rAxxAgCNAx0QDJEDKHoCIBERERMHEREpwsRB\nRESKMHEQEZEiTBxERKQIEwcRESnCxEFERIowcRARkSJMHEREpIjND3IiupNe7YqM63qLcleVhBKT\n5bAbRZLlW620NgpLTKWu88wdQ32oXVxgNBhQ7GL5nG9nUlRiRIbeYFFe2r7ydlPDW8vBEsl5MXHQ\nfSkoEThpZViQhtU9cOpygUV5I53lKMv3aqO0dZ64Y6gPtVoNo9GIxtWd+8A5v9iAYxdt31ft6ngx\ncZBTc+5PHBEROR0mDiIiUoSJg4iIFGHiICIiRZg4iIhIESYOIiJShImDiIgUYeIgIiJFmDiIiEgR\nJg4iIlKEQ448BHKLjMgtNlqUc8wjIqoITBwPgdxiI/ZZGfOJYx4RUUVg4iCZlJcL5OeiRKOFpP/f\nYIIl9k88UokeyMu1nOHuZfdYHgVSXi5KrmT99ZrfpvOG8PR2TFDk1Jg46C/5uTAe2guDxg1GffGt\nstC29o/DUAJx8bxleSmj5tIDys+F4eihv17z/1G3ag8wcZAVPDlORESKMHEQEZEiTBxERKQIEwcR\nESli15PjKSkpWL58OYQQ6NKlC3r27Gk2f8+ePdi4cSMAQKvV4rXXXkPt2rXtGSIREZXBbkccJpMJ\nCQkJmDp1KmJjY5GUlITMzEyzOtWrV8d7772HDz74AL1798ann35qr/CIiMhGdksc6enpCAgIgL+/\nP1xcXBAZGYnk5GSzOqGhoXB3dwcANGjQADk5OfYKj4iIbGS3rqqcnBz4+fnJ076+vkhPTy+1/o4d\nO9C8eXN7hPbQKjEKZFzXW5RzKBJLek0VnPENtij31XhC54B47sZhZciZOOUNgMeOHcPPP/+MGTNm\nWJ2fmpqK1NRUeTo6Oho6nTN8vO9No9FUSJzawgJoNG4W5UVGCUcuF1mUR9bxgk5neTNdiUYLg8YN\narVabk+tUkOttvxiUqlU0Ggs3z6udyxrXt9g0U5pbVirCwCSJJmVS5IKarVlubW6ZZXfMEo4dbnE\nojyyuiv87toepftEKmWflLavtBo3i9cnq7AAhy7esIyvlNdSiRKNFsJKLC4aLVyd7HNVUZ+h8lZZ\n4gSAxMRE+f+wsDCEhYWVuYzdEoevry+ys7Pl6ZycHPj6+lrUO3PmDJYtW4YpU6bAw8P6B8LaxuXn\n55dvwBVAp9NVSJxFej30d931CwAlRler5UX6YuTnC4tySV8Eo74YGo2bvJzRZITRaPlL12QyKVqn\nyU2yaKe0NqzVBQAhhFm5Wg0YjUaLcmt177fcaDJaxKh0nwij9oFfn9Je49JeSyUkfRHURivbqS9C\nkZN9rirqM1TeKlOc0dHRipez2zmOkJAQZGVl4cqVKzAYDEhKSkJERIRZnezsbMTGxmLUqFGoUaOG\nvUIjIiIF7HbEoVKpMGTIEMyaNQtCCHTt2hVBQUHYtm0bJElCVFQU1q5di4KCAiQkJEAIAbVajfff\nf99eIRIRkQ3seo6jefPmWLhwoVlZt27d5P9HjBiBESNG2DMkIiJSiHeOExGRIkwcRESkCBMHEREp\nwsRBRESKMHEQEZEiTBxERKSIUw45Qo6RK2mR4xsMtYsLjAYDAKDYxXJIjEeJtTGsHvV9QsTEQbLc\nEmDfxSKo1Wp5SI3G1R/tg9ICA3Diovl4X4/6PiHiJ4CIiBRh4iAiIkWYOIiISBEmDiIiUoQnx20g\n5eUC+bmWM3TeEJ7eDo9DUvvYLYaySCV6IM/KvnL3sqwrhM11K4PStsfo7Wq1XOWptl6eB0gF5g9t\nUpmqWq0rFWoAL43tMVp7DxUW2rw8EcDEYZv8XBgP7bUoVrdqD9gxcZQWB8K7AnCSx4caSiAunrcs\nr275UC5hMNhctzIodXse01ktF9U9rJdXU8OYtt+8LLSt9bZr6QAoSLRW3kPq+o1sX54I7KoiIiKF\nmDiIiEgRJg4iIlKE5zgqkdtDgtytWHIBIGxup8QokHFdb1FeJPHt8LDILTIit9hoUe4jaaGz4zq9\n3dTw1jrJ+TcqN/ymqERuDwlyt8aBytrJ15tw6nKBRXkjnXS/oZGTyS02Yt+Z6xbl7b1QcYnDyjrb\n1fFi4ngIsauKiIgUYeIgIiJFmDiIiEgRJg4iIlKEiYOIiBRh4iAiIkWYOIiISBEmDiIiUoSJg4iI\nFHmo7xy/XmxCVr7l0Bo1dBp4udmeM0sb6sNX0ioZ0JrIJnpNFZy56/1W7OJmva7a1erwMYUlJpvX\nl6vxxHXv2jAaDGblFfn+Lm1IlMo6RMnd26MtLECRXl9pt6csD3XiKCwx4bcL+RblXer5KEscpQz1\nEVlL0ZMQiGxSYABO3PV+a1zd+vu1oETgpJWhRRoqeKZJrlGF/RcKYTSaf5FX5Pu7tCFRKusQJXdv\nj0bjBr2+uNJuT1nYVUVERIowcRARkSJMHEREpAgTBxERKcLEQUREitj1qqqUlBQsX74cQgh06dIF\nPXv2tKjz2WefISUlBW5ubnjjjTdQt25de4ZIRERlsNsRh8lkQkJCAqZOnYrY2FgkJSUhMzPTrM7h\nw4dx6dIlfPzxxxg2bBji4+PtFR4REdnIbokjPT0dAQEB8Pf3h4uLCyIjI5GcnGxWJzk5GZ06dQIA\nNGjQADdv3kRubq69QiQiIhvYLXHk5OTAz89Pnvb19UVOTo7iOkRE5FgP9Z3jGrWEkGpVLctdeE0A\nEdH9koQQwh4rSktLw5o1azB16lQAwIYNGwDA7AT5smXL0LRpU7Rv3x4AMG7cOMTExMDb29usrdTU\nVKSmpsrT0dHRFR0+EdFDKTExUf4/LCwMYWFhZS5jt5/eISEhyMrKwpUrV2AwGJCUlISIiAizOhER\nEdi1axeAW4mmatWqFkkDuLVx0dHR8t+dG+7MGGf5YpzlpzLECDDO8paYmGj2XWpL0gDs2FWlUqkw\nZMgQzJo1C0IIdO3aFUFBQdi2bRskSUJUVBRatmyJw4cPY/To0dBqtRg5cqS9wiMiIhvZ9RxH8+bN\nsXDhQrOybt26mU0PGTLEniEREZFC6piYmBhHB1Eeqlev7ugQbMI4yxfjLD+VIUaAcZa3+4nTbifH\niYjo4cDrUomISBEmDiIiUqRS3wC4f/9+rFmzBufPn8f777+PevXqAQCOHDmCVatWwWg0wsXFBf36\n9UPTpk2dLs6CggLExsbi9OnT6Ny5MwYPHux0MQLA+vXrsXPnTqjVagwcOBDh4eEOi/NOZ86cQXx8\nPIqLi+Hv748xY8ZAq9U6OiwLGRkZiI+PR0lJCdRqNYYOHYr69es7OiwLCxYswMWLFwHcem96eHhg\n3rx5Do7Kuh9++AFbt26FSqVCy5Yt0a9fP0eHZGHNmjXYsWMHvLxuPYD3lVdeQfPmzR0cVek2bdqE\nlStXIiEhAR4eZTx6WFRimZmZ4sKFCyImJkacPn1aLv/zzz/FtWvXhBBCnD17VgwfPtxRIQohSo+z\nqKhInDx5Umzbtk0kJCQ4MMLSYzx37pyYMGGCMBgM4tKlS2LUqFHCZDI5MNK/TJo0SZw4cUIIIcTO\nnTvF6tWrHRyRdbNmzRIpKSlCCCF+++03ERMT4+CIyrZixQqxdu1aR4dh1bFjx8TMmTOFwWAQQghx\n/fp1B0dkXWJioti0aZOjw7BJdna2mDVrlnj99ddFfn5+mfUrdVdVzZo1ERAQYFFet25d+cbBWrVq\noaSkBAaDwd7hyUqL083NDQ0bNoSLi+MP/EqL8eDBg2jfvj3UajWqV6+OgIAApKenOyBCS1lZWWjU\nqBEAoFmzZjhw4ICDI7JOkiTcvHkTAHDjxg34+Pg4OKKy7du3D5GRkY4Ow6qtW7eiZ8+eUKvVAABP\nT08HR1Q6UUmuPVqxYgX69+9vc33Hf2NVsP379yM4ONgpvpwro5ycHISGhsrTzjTwZFBQEA4ePIiI\niAjs27cPV69edXRIVg0YMACzZ8/GF198AQCYOXOmgyO6txMnTsDb2xs1atRwdChWXbx4EcePH8dX\nX30FjUaDf/7zn07Z9QcAP/74I3755RfUr18fr776Ktzd3R0dkoWDBw/Cz88PtWvXtnkZp/82nTlz\nJq5fvy5PCyEgSRJefvlliyFL7nbu3DmsWrUK06ZNq+gwHyhOe6kMMd7tXjGPHDkSn3/+OdatW4dW\nrVo59MfBveI8evQoBg4ciCeeeAL79+/H0qVL8c477zhdnLffA0lJSQ4/2rhXnEajETdu3MDs2bOR\nnp6Ojz76CIsXL3a6OHv06IE+ffpAkiSsXr0aK1ascNhoGPeKc/369WbfkbYcJTl94rjfD9jVq1cx\nf/58jBo1yi434jjqi0CJ+4nR19cX2dnZ8vTVq1fh6+tbnmHdU1kx3x408+LFizh8+LA9QrLqXnEu\nXrwYgwYNAgC0bdsWS5cutVdYFsranyaTCQcOHHD4SfF7xblt2za0adMGwK0x8CRJQn5+PnQ6nb3C\nk9n6mXryyScduk9Li/Ps2bO4fPkyJkyYACEEcnJyMGnSJMyZM0c+qW9NpT7HUZqbN29i7ty56Nev\nn1k3CykXERGBvXv3wmAw4PLly8jKykJISIijwwIA5OXlAbj1Zbdu3TqL4Wucha+vL44fPw4AOHr0\nKGrWrOngiEp35MgRBAUF2fXHgVKtW7fGsWPHAAAXLlyA0Wh0SNIoy50PoTtw4ABq1arlwGisq127\nNuLj47F48WIsWbIEvr6+mDdv3j2TBlDJ7xz/9ddf8fnnnyMvLw9Vq1ZF3bp1MWXKFHzzzTfYsGED\nAgIC5EOyqVOnOuwkWmlxAsAbb7yBoqIiGAwGuLu7Y9q0aQgMDHSqGNevX4+ffvoJLi4uTnU57ubN\nm7FlyxZIkoQnnngC//jHPxwdklWnTp3C559/DpPJBFdXVwwdOhTBwcGODsuquLg4hIaGIioqytGh\nlMpgMGDp0qXIyMiAq6srXn31VTRp0sTRYVlYvHgxMjIyIEkS/P39MWzYMKujfTuTUaNGYe7cuWVe\njlupEwcREdnfQ9lVRUREFYeJg4iIFGHiICIiRZg4iIhIESYOIiJShImDiIgUYeIgIiJFmDiIiEiR\n/weq+jQQ+NlsNgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x12b279e8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "motif,nonMotif,motifName=seperateMotifAndNonMotif('P4__G',data,'TC')\n",
    "histOfMotfAndNonMotif(motif,nonMotif,motifName)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "More data about the motifs that could give an idea about other factors influencing the substitution rates other than the motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "def parseSeq(seqLocation,codingStart,codingEnd):\n",
    "    fasta_sequences = SeqIO.parse(open(seqLocation),'fasta')\n",
    "    for fasta in fasta_sequences:\n",
    "        consensus= fasta.seq[codingStart:codingEnd]\n",
    "        return consensus\n",
    "\n",
    "def isSynMutation(seq,kmer,offset,k=5):\n",
    "    originalSeq=str(seq)\n",
    "    mutatedSeq=str(seq)  \n",
    "    mutatedSeqList=list(mutatedSeq)\n",
    "    mutatedSeqList[offset-k//2:offset+k//2+1]=list(kmer)\n",
    "    mutatedSeq=\"\".join(mutatedSeqList)\n",
    "    return translate(originalSeq)==translate(mutatedSeq)\n",
    "    \n",
    "def checkMotiflocations(seq,motif,mutationType):\n",
    "    locations=[]\n",
    "    motif=featureToMotiff(motif,mutationType,toString=False)\n",
    "    k=len(motif)\n",
    "    #print motif\n",
    "    allIndices=set([i for i in range(k)])\n",
    "    freeIndices=set([i for i, x in enumerate(motif) if x == \"X\"])\n",
    "    #print freeIndices\n",
    "    checkIndices=list(allIndices-freeIndices)\n",
    "    #print checkIndices\n",
    "    for i in range(k//2,len(seq)-k//2):\n",
    "        isMotif=True\n",
    "        compared=list(seq[i-k//2:i+k//2+1])\n",
    "        #print compared\n",
    "        for position in checkIndices:\n",
    "            #print (compared[position],motif[position])\n",
    "            if compared[position]!=motif[position]:\n",
    "                isMotif=False\n",
    "                break\n",
    "        comparedAfterMutation=compared\n",
    "        comparedAfterMutation[k//2]=mutationType[1]\n",
    "        if (isMotif and isSynMutation(seq,comparedAfterMutation,i)):\n",
    "            locations.append(i)\n",
    "    return locations\n",
    "\n",
    "def is4FoldDegenerate(seq,location,k=5):\n",
    "    compared=list(seq[location-k//2:location+k//2+1])\n",
    "    for nuc in NUCS:\n",
    "        compared[k//2]=nuc\n",
    "        if not(isSynMutation(seq,compared,location)):\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "\n",
    "def iterateOverLocationsList(function,seq,locations,k=5):\n",
    "    resList=[]\n",
    "    for location in locations:\n",
    "        if function(seq,location):\n",
    "            resList.append(location)\n",
    "    return resList\n",
    "            \n",
    "def hyperGeoTest(M,n,locations,TrueLocations):\n",
    "    rv = hypergeom(M, n, locations)\n",
    "    return rv.pmf(TrueLocations)\n",
    "\n",
    "def checkHyperGeometricForAllMotifs(motifTable,function,M,n,seq,thrashold=0.05):\n",
    "    sigHG={}\n",
    "    mutations=motifTable.columns\n",
    "    for mt in mutations:\n",
    "        lenn=len(motifTable[mt])\n",
    "        tableByMt=motifTable[mt]\n",
    "        for i in range(lenn):\n",
    "            cellMotifs=tableByMt[i]\n",
    "            if cellMotifs!=\"\":\n",
    "                for motif in cellMotifs:\n",
    "                    #print motif\n",
    "                    allLocations=checkMotiflocations(seq,motif,mt)\n",
    "                    #print len(allLocations)\n",
    "                    trueLocations=iterateOverLocationsList(function,seq,allLocations)\n",
    "                    #print len(trueLocations)\n",
    "                    pvalue=hyperGeoTest(M,n,len(allLocations),len(trueLocations))\n",
    "                    #print pvalue\n",
    "                    if (featureToMotiff(motif,mt),mt) in sigHG.keys():\n",
    "                        continue\n",
    "                    if (pvalue < thrashold):\n",
    "                        sigHG[(featureToMotiff(motif,mt),mt)]=pvalue\n",
    "    return pd.DataFrame(sigHG.items(),columns=['motif_mutationType','HG test PV'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>4-fold Degenerate sites enrichment</h3>\n",
    "\n",
    "We want to check the if there is any kind of enrichment for 4-fold degenerate site. in order to do this we will preform a HG test when $n=994$ is the number of 4-fold degenerate sites within the coding region and $M=2345$ possible sites of synonymous mutation.\n",
    "\n",
    "<b>There appears to be an significant large number of 4-fold degenerate sites in many of the motif we found as having an effect on mutation rate</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 389,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_mutationType</th>\n",
       "      <th>HG test PV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(AGGXC, GA)</td>\n",
       "      <td>2.092757e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(GXTTG, TA)</td>\n",
       "      <td>3.217076e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(AXATA, AC)</td>\n",
       "      <td>3.217076e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(XXTAX, TC)</td>\n",
       "      <td>1.274725e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(GCCXX, CG)</td>\n",
       "      <td>7.755594e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>(XCAXX, AC)</td>\n",
       "      <td>6.681771e-53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>(CAATX, AG)</td>\n",
       "      <td>3.639374e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>(XXATG, AT)</td>\n",
       "      <td>4.712874e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>(XXTAA, TC)</td>\n",
       "      <td>4.990953e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>(XTAGX, AG)</td>\n",
       "      <td>3.247597e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>(XXTCC, TG)</td>\n",
       "      <td>1.362744e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>(XCTXG, TA)</td>\n",
       "      <td>1.220260e-11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   motif_mutationType    HG test PV\n",
       "0         (AGGXC, GA)  2.092757e-02\n",
       "1         (GXTTG, TA)  3.217076e-02\n",
       "2         (AXATA, AC)  3.217076e-02\n",
       "3         (XXTAX, TC)  1.274725e-02\n",
       "4         (GCCXX, CG)  7.755594e-16\n",
       "5         (XCAXX, AC)  6.681771e-53\n",
       "6         (CAATX, AG)  3.639374e-02\n",
       "7         (XXATG, AT)  4.712874e-02\n",
       "8         (XXTAA, TC)  4.990953e-02\n",
       "9         (XTAGX, AG)  3.247597e-02\n",
       "10        (XXTCC, TG)  1.362744e-05\n",
       "11        (XCTXG, TA)  1.220260e-11"
      ]
     },
     "execution_count": 389,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cons=parseSeq(r'C:\\Users\\Guyling1\\ContextProject\\AggarwalaPaper\\sabin2.full_genome.U882C.A2973G.C4905U.C5526U.fasta',747,7638)\n",
    "checkHyperGeometricForAllMotifs(rawMotifSummary,is4FoldDegenerate,2345,994,cons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
